File: lib/src/main/java/com/asyncanimator/launcher/seq/AnimationSeqHelper.java
package com.asyncanimator.launcher.seq;

import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import com.asyncanimator.util.Trace;

import java.util.HashMap;
import java.util.Map;

/**
 * AnimationSeqHelper — Recents 动画 SeqId 防抖。
 *
 * <p>对应分析文档 §6.9。
 *
 * <p>核心机制：
 * <ul>
 *   <li>全局单调 seqId 写入 Bundle（跨进程同步）</li>
 *   <li>{@code canFinishRecent()} 检查 500ms 内是否刚结束过</li>
 *   <li>{@code canInterceptGesture()} 检查 300ms 内是否刚启动过 app</li>
 *   <li>{@code delayFinishRecents(Runnable)} 500ms 内必须 finish</li>
 * </ul>
 */
public class AnimationSeqHelper extends DefaultAnimationSeqHelper {

    public static final long MAX_DELAY_TIME = 500L;
    public static final long MAX_GO_NORMAL_DELAY_TIME = 200L;
    public static final long MAX_INTERCEPT_GESTURE_DELAY_TIME = 300L;

    private static final int MSG_EXC_RUNNABLE = 1;
    private static final String KEY_INTERRUPT_TRANSITION_START_ACTIVITY_SEQ_ID =
        "interrupt.transition.startActivity.seqId";

    private long seqId = 0;
    private Runnable delayRunnable;
    /** 懒创建：JVM 单测环境没有主 Looper，构造期不触碰 android.os.Handler。 */
    private Handler handler;

    private Handler getOrCreateHandler() {
        if (handler == null) {
            handler = new Handler(Looper.getMainLooper(), msg -> {
                if (msg.what == MSG_EXC_RUNNABLE) {
                    Trace.traceBegin(8L, "exc delayRunnable");
                    if (delayRunnable != null) delayRunnable.run();
                    delayRunnable = null;
                    Trace.traceEnd(8L);
                }
                return true;
            });
        }
        return handler;
    }

    /** (controller, seqId) Pair，记录当前 recents controller 对应的 seqId。 */
    private Object nextFinishSeqId;

    private long updateSeqId() {
        return ++seqId;
    }

    @Override
    public void addSeqId(Bundle bundle) {
        if (bundle == null) return;
        long id = updateSeqId();
        bundle.putLong(KEY_INTERRUPT_TRANSITION_START_ACTIVITY_SEQ_ID, id);
    }

    @Override
    public boolean canFinishRecent() {
        return AnimSeqTimeStamp.getTimeGapToLastRecentFinishTime() > MAX_DELAY_TIME;
    }

    @Override
    public boolean canInterceptGesture() {
        return AnimSeqTimeStamp.getTimeGapToLastStartAppTime() > MAX_INTERCEPT_GESTURE_DELAY_TIME;
    }

    @Override
    public boolean delayFinishRecents(Runnable r) {
        if (canFinishRecent()) {
            if (r != null) r.run();
            return false;
        }
        Trace.traceBegin(8L, "delayFinishRecents");
        clearFinishRecentsRunnable();
        delayRunnable = r;
        long delay = MAX_DELAY_TIME - AnimSeqTimeStamp.getTimeGapToLastRecentFinishTime();
        getOrCreateHandler().sendEmptyMessageDelayed(MSG_EXC_RUNNABLE, Math.max(0L, delay));
        Trace.traceEnd(8L);
        return true;
    }

    @Override
    public void clearFinishRecentsRunnable() {
        if (handler != null) handler.removeMessages(MSG_EXC_RUNNABLE);
        delayRunnable = null;
    }

    @Override
    public void updateNextFinishSeqIdIfNeed(Object recentsController) {
        long id = updateSeqId();
        nextFinishSeqId = new SeqIdPair(recentsController, id);
    }

    @Override
    public long getNextFinishSeqId(Object recentsController) {
        if (nextFinishSeqId instanceof SeqIdPair) {
            SeqIdPair p = (SeqIdPair) nextFinishSeqId;
            if (p.controller == recentsController) return p.seqId;
        }
        return 0L;
    }

    /** 简化的 Pair（避免 Kotlin Pair 依赖）。 */
    private static class SeqIdPair {
        final Object controller;
        final long seqId;
        SeqIdPair(Object c, long s) { this.controller = c; this.seqId = s; }
    }
}

File: lib/src/main/java/com/asyncanimator/launcher/seq/AnimationSeqHelper.kt
package com.asyncanimator.launcher.seq

import android.os.Bundle
import android.os.Handler
import android.os.Looper
import com.asyncanimator.util.Trace

internal const val MAX_DELAY_TIME = 500L
internal const val MAX_INTERCEPT_GESTURE_DELAY_TIME = 300L

private const val MSG_EXC_RUNNABLE = 1
private const val KEY_INTERRUPT_TRANSITION_START_ACTIVITY_SEQ_ID =
    "interrupt.transition.startActivity.seqId"

/**
 * AnimationSeqHelper — Recents 动画 SeqId 防抖。
 *
 * 对应 `docs/review/03-controller-manager-seq.md`。
 *
 * 核心机制：
 *
 *  - 全局单调 seqId 写入 Bundle（跨进程同步）
 *  - [canFinishRecent] 检查 500ms 内是否刚结束过
 *  - [canInterceptGesture] 检查 300ms 内是否刚启动过 app
 *  - [delayFinishRecents] 500ms 内必须 finish
 */
class AnimationSeqHelper : DefaultAnimationSeqHelper() {

    private var seqId = 0L
    private var delayAction: (() -> Unit)? = null

    /** 懒创建：JVM 单测环境没有主 Looper，构造期不触碰 android.os.Handler。 */
    private var handler: Handler? = null

    /** (controller, seqId) Pair，记录当前 recents controller 对应的 seqId。 */
    private var nextFinishSeqId: Pair<Any?, Long>? = null

    private fun getOrCreateHandler(): Handler {
        if (handler == null) {
            handler = Handler(Looper.getMainLooper()) { msg ->
                if (msg.what == MSG_EXC_RUNNABLE) {
                    Trace.traceBegin(8L, "exc delayRunnable")
                    delayAction?.invoke()
                    delayAction = null
                    Trace.traceEnd(8L)
                }
                true
            }
        }
        return handler!!
    }

    private fun updateSeqId(): Long = ++seqId

    override fun addSeqId(bundle: Bundle?) {
        if (bundle == null) return
        val id = updateSeqId()
        bundle.putLong(KEY_INTERRUPT_TRANSITION_START_ACTIVITY_SEQ_ID, id)
    }

    override val canFinishRecent: Boolean
        get() = AnimSeqTimeStamp.timeGapToLastRecentFinishTime > MAX_DELAY_TIME

    override val canInterceptGesture: Boolean
        get() = AnimSeqTimeStamp.timeGapToLastStartAppTime > MAX_INTERCEPT_GESTURE_DELAY_TIME

    override fun delayFinishRecents(action: (() -> Unit)?): Boolean {
        if (canFinishRecent) {
            action?.invoke()
            return false
        }
        Trace.traceBegin(8L, "delayFinishRecents")
        clearFinishRecentsRunnable()
        delayAction = action
        val delay = MAX_DELAY_TIME - AnimSeqTimeStamp.timeGapToLastRecentFinishTime
        getOrCreateHandler().sendEmptyMessageDelayed(MSG_EXC_RUNNABLE, maxOf(0L, delay))
        Trace.traceEnd(8L)
        return true
    }

    override fun clearFinishRecentsRunnable() {
        handler?.removeMessages(MSG_EXC_RUNNABLE)
        delayAction = null
    }

    override fun resetInterceptState() {
        AnimSeqTimeStamp.resetLastStartAppTime()
    }

    override fun updateNextFinishSeqIdIfNeed(recentsController: Any?) {
        val p = nextFinishSeqId
        // 原厂仅在 pair 为空或 controller 变更时才更新（AnimationSeqHelper.java:123-127）
        if (p == null || p.first != recentsController) {
            nextFinishSeqId = recentsController to updateSeqId()
        }
    }

    override fun getNextFinishSeqId(recentsController: Any?): Long {
        val p = nextFinishSeqId
        // 原厂按引用比较 controller，这里用 === 保持一致
        if (p != null && p.first == recentsController) return p.second
        return 0L
    }
}
